File: src/2022/09.ts
// https://adventofcode.com/2022/day/9

interface Knot {
  x: number;
  y: number;
}

export class Solver {
  private _input: string[];
  private _rope: Knot[];
  private _tailPositionSet: Set<string>;

  constructor(input: string[]) {
    this._input = input;
  }

  private _createRope(length: number): void {
    this._rope = [];

    for (let i = 0; i < length; i++) {
      this._rope.push({ x: 0, y: 0 });
    }
  }

  private _moveHead(direction: string): void {
    const head = this._rope[0];

    switch (direction) {
      case "U":
        head.y -= 1;
        break;
      case "R":
        head.x += 1;
        break;
      case "D":
        head.y += 1;
        break;
      case "L":
        head.x -= 1;
        break;
    }
  }

  private _move(direction: string, distance: number): void {
    for (let i = 0; i < distance; i++) {
      // Move head knot one position
      this._moveHead(direction);

      // Move each subsequent knot if needed
      for (let j = 1; j < this._rope.length; j++) {
        const current = this._rope[j];
        const previous = this._rope[j - 1];

        const horizontalDistance = previous.x - current.x;
        const verticalDistance = previous.y - current.y;

        if (
          Math.abs(horizontalDistance) >= 1 &&
          Math.abs(verticalDistance) >= 1 &&
          (Math.abs(horizontalDistance) > 1 || Math.abs(verticalDistance) > 1)
        ) {
          // Move diagonally
          current.x += Math.sign(horizontalDistance);
          current.y += Math.sign(verticalDistance);
        } else if (Math.abs(horizontalDistance) > 1) {
          // Move horizontally
          current.x += Math.sign(horizontalDistance);
        } else if (Math.abs(verticalDistance) > 1) {
          // Move vertially
          current.y += Math.sign(verticalDistance);
        }

        // This is the tail (the very last knot of the rope)
        if (j === this._rope.length - 1) {
          const { x, y } = current;
          this._tailPositionSet.add(`${x},${y}`);
        }
      }
    }
  }

  private _executeMovements(): void {
    this._tailPositionSet = new Set<string>();

    this._input.forEach(line => {
      const match = line.match(/(\w) (\d+)/);
      const direction = match[1];
      const distance = parseInt(match[2]);

      this._move(direction, distance);
    });
  }

  public part1(): number {
    this._createRope(2);
    this._executeMovements();

    return this._tailPositionSet.size;
  }

  public part2(): number {
    this._createRope(10);
    this._executeMovements();

    return this._tailPositionSet.size;
  }
}

File: rust/2022/src/bin/09.rs
// https://adventofcode.com/2022/day/9

use regex::Regex;
use std::collections::HashSet;

#[derive(Clone, Default, Eq, Hash, PartialEq)]
struct Point {
    x: i16,
    y: i16,
}

fn move_rope(
    rope: &mut Vec<Point>,
    tail_position_set: &mut HashSet<Point>,
    direction: char,
    distance: u8,
) {
    for _ in 0..distance {
        // Move head knot one position
        match direction {
            'U' => rope[0].y -= 1,
            'R' => rope[0].x += 1,
            'D' => rope[0].y += 1,
            'L' => rope[0].x -= 1,
            _ => unreachable!(),
        }

        let rope_length = rope.len();

        // Move each subsequent knot if needed
        for j in 1..rope_length {
            let previous = rope[j - 1].clone();
            let current = &mut rope[j];

            let horizontal_distance = previous.x - current.x;
            let vertical_distance = previous.y - current.y;

            if horizontal_distance == 0 && vertical_distance == 0 {
                // We're all caught up, so no further movement is needed
                break;
            }

            if horizontal_distance.abs() > 1 || vertical_distance.abs() > 1 {
                // Move diagonally
                current.x += horizontal_distance.signum();
                current.y += vertical_distance.signum();
            } else if horizontal_distance.abs() > 1 {
                // Move horizontally
                current.x += horizontal_distance.signum();
            } else if vertical_distance.abs() > 1 {
                // Move vertially
                current.y += vertical_distance.signum();
            }

            // This is the tail knot
            if j == rope_length - 1 {
                tail_position_set.insert(current.clone());
            }
        }
    }
}

fn count_tail_positions(length: usize, input: &str) -> usize {
    let re = Regex::new(r"(\w) (\d+)").unwrap();
    let mut rope = vec![Point::default(); length];
    let mut tail_position_set = HashSet::new();

    input.lines().for_each(|line| {
        let caps = re.captures(line).unwrap();
        let direction = caps[1].chars().nth(0).unwrap();
        let distance = caps[2].parse::<u8>().unwrap();

        move_rope(&mut rope, &mut tail_position_set, direction, distance);
    });

    tail_position_set.len()
}

fn part1(input: &str) -> usize {
    count_tail_positions(2, input)
}

fn part2(input: &str) -> usize {
    count_tail_positions(10, input)
}

fn main() {
    let input = include_str!("../../../../_input/2022/09.txt").trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2: {}", part2(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../../../_input/2022/09.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 5981);
    }

    #[test]
    fn part2() {
        assert_eq!(super::part2(INPUT.trim_end()), 2352);
    }
}
